src/probability_theory/density.lean

```diff
@@ -0,0 +1,380 @@
+
+include hX
+
+/-- A real-valued random variable `X` `has_pdf X ℙ λ` (where `λ` is the Lebesgue measure) if and
+only if the push-forward measure of `ℙ` along `X` is absolutely continuous with respect to `λ`. -/
+lemma has_pdf_iff : has_pdf X ℙ ↔ map X ℙ ≪ volume :=
+begin
+  split,
+  { intro hX',
+    exactI map_absolutely_continuous hX },
+  { intro h,
+    refine ⟨⟨(measure.map X ℙ).radon_nikodym_deriv volume, measurable_radon_nikodym_deriv _ _, _⟩⟩,
+    haveI := ℙ.is_finite_measure_map hX,
+    rwa with_density_radon_nikodym_deriv_eq }
+end
+
+/-- A real-valued random variable that `has_pdf` transformed under a `quasi_measure_preserving`
+map also `has_pdf`. -/
+lemma quasi_measure_preserving_has_pdf [has_pdf X ℙ]
+  (g : ℝ → ℝ) (hg : quasi_measure_preserving g) :
+  has_pdf (g ∘ X) ℙ :=
+begin
+  rw [has_pdf_iff (hg.measurable.comp hX), ← map_map hg.measurable hX, pdf_spec X ℙ],
+  refine absolutely_continuous.mk (λ s hsm hs, _),
+  rw [map_apply hg.measurable hsm, with_density_apply _ (hg.measurable hsm)],
+  have := hg.absolutely_continuous hs,
+  rw map_apply hg.measurable hsm at this,
+  exact set_lintegral_measure_zero _ _ this,
+  apply_instance
+end
+
+/-- If `X` is a real-valued random variable that has pdf `f`, then the expectation of `X` equals
+`∫ x, x * f x ∂λ` where `λ` is the Lebesgue measure. -/
+lemma integral_mul_eq_integral [has_pdf X ℙ]
+  (hpdf : integrable (λ x, x * (pdf X ℙ volume x).to_real) volume) :
+  ∫ x, x * (pdf X ℙ volume x).to_real ∂(volume) = ∫ x, X x ∂ℙ :=
+integral_mul_eq_integral' hX _ measurable_id hpdf
+
+lemma has_finite_integral_mul {f : ℝ → ℝ} {g : ℝ → ℝ≥0∞}
+  (hg : pdf X ℙ =ᵐ[volume] g) (hgi : ∫⁻ x, ∥f x∥₊ * g x ∂(volume) < ∞) :
+  has_finite_integral (λ x, f x * (pdf X ℙ volume x).to_real) volume :=
+begin
+  rw [has_finite_integral],
+  have : (λ x, ↑∥f x∥₊ * g x) =ᵐ[volume] (λ x, ∥f x * (pdf X ℙ volume x).to_real∥₊),
+  { refine ae_eq_trans (filter.eventually_eq.mul (ae_eq_refl (λ x, ∥f x∥₊))
+      (ae_eq_trans hg.symm (of_real_to_real_ae_eq hX).symm)) _,
+    simp_rw [← smul_eq_mul, nnnorm_smul, ennreal.coe_mul, smul_eq_mul],
+    refine filter.eventually_eq.mul (ae_eq_refl _) _,
+    convert ae_eq_refl _,
+    ext1 x,
+    exact real.ennnorm_eq_of_real ennreal.to_real_nonneg },
+  rwa ← lintegral_congr_ae this,
+end
+
+end real
+
+section
+
+/-! **Uniform Distribution** -/
+
+/-- A random variable `X` has uniform distribution if it has a probability density function `f`
+with compact support `s` such that `f = (μ s)⁻¹ 1ₛ` a.e. where `1ₛ` is the indicator function
+for `s`.
+
+We require compact support since otherwise the expectation might not be finite. -/
+class uniform {m : measurable_space α} (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac)
+  extends has_pdf X ℙ μ :=
+(support' : set E) (compact_support' : is_compact support')
+(support_not_null' : 0 < μ support')
+(uniform' : pdf X ℙ μ =ᵐ[μ] support'.indicator ((μ support')⁻¹ • 1))
```

Comment:
I'm a bit surprised that you are bundling `support` as a field, rather than having it as a parameter. Is it really useful to know that a random variable is uniform on some unknown set, rather than knowing it has uniform distribution on a particular stated set?

Reply:
I originally had the version with the support being a field but in a meeting with @RemyDegenne , he suggested it might be useful to talk about a family of distributions. I'm not quite sure which is better

Reply:
Could either you or Remy explain the application to families? I'm not seeing it (but I'm not a probabilist).

Reply:
I think I indeed said this, but now it does not look wise. The support may be better as a parameter.
With the support as a parameter, there is another linked question I am unsure about: why is `uniform` a class, and not a Pop valued def?
```
def uniform {m : measurable_space α} (s : set E) (X : α → E)
  (ℙ : measure α) (μ : measure E . volume_tac) [has_pdf X ℙ μ] : 
  Prop :=
pdf X ℙ μ =ᵐ[μ] s.indicator ((μ s)⁻¹ • 1)
```
The common reason to make it a class would be that we can deduce `uniform` from other instances, and I don't think that will happen.

Reply:
My rationale for making it a class is that we won't have more than 1 uniform distribution on each random variable and so, making it into a class allows us to not carry a prop around.  Though perhaps its not good enough of a reason.

I will split the uniform stuff into another PR since it doesn't have much to do with the pdf stuff

Reply:
#9385

Reply:
> The common reason to make it a class would be that we can deduce `uniform` from other instances, and I don't think that will happen.

Another reason for why we might want class over def is that, we have all exponential distributions are gamma distributions, so we can infer gamma from exponential

Reply:
How about we make it a `def` for now, and it can potentially become a `class` later when we have a concrete example of this being useful?